tests/drivers/adc/adc_emul/src/main.c

```diff
@@ -0,0 +1,625 @@
+
+	/* Test sampling */
+	start_adc_read(adc_dev, BIT(ADC_1ST_CHANNEL_ID), samples);
+
+	/*
+	 * Check samples - returned value should max out on reference value.
+	 * Raw value shoudn't exceed resolution.
+	 */
+	check_samples(samples, ADC_REF_INTERNAL_MV, 0 /* step */,
+		      1 /* channels */, 0 /* first channel data */,
+		      ADC_REF_INTERNAL_MV, ADC_GAIN_1);
+
+	check_empty_samples(samples);
+
+	for (i = 0; i < samples; i++)
+		zassert_equal(BIT_MASK(ADC_RESOLUTION), m_sample_buffer[i],
+			      "[%u] raw value isn't max value", i);
+}
+
+/*
+ * Test different reference sources and if error is reported when unconfigured
+ * reference source is requested.
+ */
+static void test_adc_emul_reference(void)
+{
+	const uint16_t input1_mv = 4000;
+	const uint16_t input2_mv = 2000;
+	const int samples = 3;
+	int ret, i;
+
+	for (i = 0; i < BUFFER_SIZE; ++i) {
+		m_sample_buffer[i] = INVALID_ADC_VALUE;
+	}
+
+	/* Generic ADC setup */
+	const struct device *adc_dev = device_get_binding(ADC_DEVICE_NAME);
+	zassert_not_null(adc_dev, "Cannot get ADC device");
+
+	channel_setup(adc_dev, ADC_REF_EXTERNAL1, ADC_GAIN_1,
+		      ADC_1ST_CHANNEL_ID);
+
+	struct adc_channel_cfg channel_cfg = {
+		.gain             = ADC_GAIN_1,
+		/* Reference value not setup in DTS */
+		.reference        = ADC_REF_EXTERNAL0,
+		.acquisition_time = ADC_ACQUISITION_TIME,
+		.channel_id       = ADC_2ND_CHANNEL_ID,
+	};
+
+	ret = adc_channel_setup(adc_dev, &channel_cfg);
+	zassert_not_equal(ret, 0,
+			  "Setting up of the %d channel shuldn't successed",
+			  ADC_2ND_CHANNEL_ID);
+
+	channel_setup(adc_dev, ADC_REF_INTERNAL, ADC_GAIN_1,
+		      ADC_2ND_CHANNEL_ID);
+
+	/* ADC emulator-specific setup */
+	ret = adc_emul_const_value_set(adc_dev, ADC_1ST_CHANNEL_ID, input1_mv);
+	zassert_ok(ret, "adc_emul_const_value_set() failed with code %d", ret);
+
+	ret = adc_emul_const_value_set(adc_dev, ADC_2ND_CHANNEL_ID, input2_mv);
+	zassert_ok(ret, "adc_emul_const_value_set() failed with code %d", ret);
+
+	/* Test sampling */
+	start_adc_read(adc_dev, BIT(ADC_1ST_CHANNEL_ID) |
+			BIT(ADC_2ND_CHANNEL_ID), samples);
+
+	/* Check samples */
+	check_samples(samples * 2, input1_mv, 0 /* step */, 2 /* channels */,
+		      0 /* first channel data */, ADC_REF_EXTERNAL1_MV,
+		      ADC_GAIN_1);
+	check_samples(samples * 2, input2_mv, 0 /* step */, 2 /* channels */,
+		      1 /* first channel data */, ADC_REF_INTERNAL_MV,
+		      ADC_GAIN_1);
+
+	check_empty_samples(samples * 2);
+}
+
+/*
+ * Test setting reference value.
+ */
+static void test_adc_emul_ref_voltage_set(void)
+{
+	const uint16_t input1_mv = 4000;
+	const uint16_t input2_mv = 2000;
+	const uint16_t ref1_mv = 6000;
+	const uint16_t ref2_mv = 9000;
+	const int samples = 3;
+	int ret, i;
+
+	for (i = 0; i < BUFFER_SIZE; ++i) {
+		m_sample_buffer[i] = INVALID_ADC_VALUE;
+	}
+
+	/* Generic ADC setup */
+	const struct device *adc_dev = device_get_binding(ADC_DEVICE_NAME);
+	zassert_not_null(adc_dev, "Cannot get ADC device");
```

Comment:
Would it be worth putting this common code in a function, to try to reduce the duplicate in this file?

Reply:
Now function `get_adc_device()` is used to obtain ADC device pointer and check if it is not null